userdb/memory/mem.go

```diff
@@ -0,0 +1,96 @@
+package memory
+
+import (
+	"bufio"
+	"encoding/csv"
+	"io"
+	"os"
+	"sync"
+
+	"github.com/janekolszak/idp/core"
+	"golang.org/x/crypto/bcrypt"
+)
+
+type Store struct {
+	hashes map[string]string
+	mtx    sync.RWMutex
+}
+
+func NewMemStore() (*Store, error) {
+	s := Store{}
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+
+	return &s, nil
+}
+
+func (s *Store) LoadHtpasswd(filename string) error {
+
+	f, err := os.OpenFile(filename, os.O_RDONLY, os.ModeExclusive)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	r := csv.NewReader(bufio.NewReader(f))
+	r.Comment = '#'
+	r.Comma = ':'
+	r.TrimLeadingSpace = true
+	r.FieldsPerRecord = 2
+
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	s.hashes = make(map[string]string)
+	for {
+		fields, err := r.Read()
+		if err != nil {
+			if err == io.EOF {
+				return nil
+			}
+
+			return err
+		}
+
+		s.hashes[fields[0]] = fields[1]
+	}
+}
+
+func (s *Store) Check(username, password string) error {
```

Comment:
Htpasswd is supposed to read htpasswd files created by the apache htpasswd tool.

Your implementation is similar, but I **please don't combine those two**. Some day htpasswd will use your implementation inside, but later.